File: Backend/htbp_parent/service/service_hosp/src/main/java/com/seclib/htbp/hosp/service/impl/DepartmentServiceImpl.java
package com.seclib.htbp.hosp.service.impl;

import com.alibaba.fastjson.JSONObject;
import com.seclib.htbp.hosp.repository.DepartmentRepository;
import com.seclib.htbp.hosp.service.DepartmentService;
import com.seclib.htbp.model.hosp.Department;
import com.seclib.htbp.vo.hosp.DepartmentQueryVo;
import com.seclib.htbp.vo.hosp.DepartmentVo;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class DepartmentServiceImpl implements DepartmentService {
    @Autowired
    private DepartmentRepository departmentRepository;

    @Override
    public void save(Map<String, Object> paramMap) {
        String paramString = JSONObject.toJSONString(paramMap);
        Department department = JSONObject.parseObject(paramString,Department.class);

        Department departmentExist = departmentRepository.getDeparmentByHoscodeAndDepcode(department.getHoscode(),department.getDepcode());

        if(departmentExist != null){
            departmentExist.setUpdateTime(new Date());
            departmentExist.setIsDeleted(0);
            departmentRepository.save(departmentExist);
        } else {
            department.setCreateTime(new Date());
            department.setUpdateTime( new Date());
            department.setIsDeleted(0);
            departmentRepository.save(department);
        }
    }

    @Override
    public Page<Department> findPageDepartment(Integer page, Integer limit, DepartmentQueryVo departmentQueryVo) {
        Pageable pageable = PageRequest.of(page -1 ,limit);
        ExampleMatcher matcher = ExampleMatcher.matching()
                .withStringMatcher(ExampleMatcher.StringMatcher.CONTAINING)
                .withIgnoreCase(true);
        Department department = new Department();
        BeanUtils.copyProperties(departmentQueryVo,department);
        Example<Department> example = Example.of(department,matcher);
        Page<Department> all =  departmentRepository.findAll(example,pageable);
        return all;
    }

    @Override
    public void remove(String hoscode, String depcode) {
        Department department = departmentRepository.getDeparmentByHoscodeAndDepcode(hoscode,depcode);
        if(department != null) {
            departmentRepository.deleteById(department.getId());
        }
    }

    @Override
    public List<DepartmentVo> findDeptTree(String hoscode) {
        List<DepartmentVo> result = new ArrayList<>();

        Department departmentQuery = new Department();
        departmentQuery.setHoscode(hoscode);

        Example<Department> example = Example.of(departmentQuery);
        List<Department> departmentList = departmentRepository.findAll(example);

        Map<String, List<Department>> departmentMap = departmentList.stream().collect(Collectors.groupingBy(Department::getBigcode));
        for(Map.Entry<String,List<Department>> entry: departmentMap.entrySet()){
            String bigcode = entry.getKey();
            List<Department> departmentByBigCodeList = entry.getValue();
            DepartmentVo departmentVo = new DepartmentVo();
            departmentVo.setDepcode(bigcode);
            departmentVo.setDepname(departmentByBigCodeList.get(0).getBigname());

            List<DepartmentVo> children = new ArrayList<>();
            for(Department department: departmentByBigCodeList){
                DepartmentVo childDepartmentVo = new DepartmentVo();
                childDepartmentVo.setDepcode(department.getDepcode());
                childDepartmentVo.setDepname(department.getDepname());
                children.add(childDepartmentVo);
            }
            departmentVo.setChildren(children);
            result.add(departmentVo);
        }

        return result;
    }
}

File: Backend/htbp_parent/service/service_hosp/src/main/java/com/seclib/htbp/hosp/service/impl/DepartmentServiceImpl.kt
package com.seclib.htbp.hosp.service.impl

import com.alibaba.fastjson.JSONObject

import com.seclib.htbp.hosp.service.DepartmentService
import org.springframework.beans.factory.annotation.Autowired
import com.seclib.htbp.hosp.repository.DepartmentRepository
import com.seclib.htbp.model.hosp.Department
import com.seclib.htbp.vo.hosp.DepartmentQueryVo
import com.seclib.htbp.vo.hosp.DepartmentVo
import java.util.stream.Collectors
import org.springframework.beans.BeanUtils
import org.springframework.data.domain.*
import org.springframework.stereotype.Service
import java.util.*
import java.util.function.Function

@Service
open class DepartmentServiceImpl : DepartmentService {
    @Autowired
    private val departmentRepository: DepartmentRepository? = null
    override fun save(paramMap: Map<String?, Any?>?) {
        val paramString = JSONObject.toJSONString(paramMap)
        val department = JSONObject.parseObject(paramString, Department::class.java)
        val departmentExist =
            departmentRepository!!.getDepartmentByHoscodeAndDepcode(department.hoscode, department.depcode)
        if (departmentExist != null) {
            departmentExist.updateTime = Date()
            departmentExist.isDeleted = 0
            departmentRepository.save(departmentExist)
        } else {
            department.createTime = Date()
            department.updateTime = Date()
            department.isDeleted = 0
            departmentRepository.save(department)
        }
    }

    override fun findPageDepartment(
        page: Int?,
        limit: Int?,
        departmentQueryVo: DepartmentQueryVo?
    ): Page<Department?>? {
        val pageable: Pageable = PageRequest.of(page!! - 1, limit!!)
        val matcher = ExampleMatcher.matching()
            .withStringMatcher(ExampleMatcher.StringMatcher.CONTAINING)
            .withIgnoreCase(true)
        val department = Department()
        BeanUtils.copyProperties(departmentQueryVo, department)
        val example =
            Example.of(department, matcher)
        return departmentRepository!!.findAll(example, pageable)
    }

    override fun remove(hoscode: String?, depcode: String?) {
        val department = departmentRepository!!.getDepartmentByHoscodeAndDepcode(hoscode, depcode)
        if (department != null) {
            departmentRepository.deleteById(department.id)
        }
    }

    override fun findDeptTree(hoscode: String?): List<DepartmentVo?>? {
        val result: MutableList<DepartmentVo?> = ArrayList()
        val departmentQuery = Department()
        departmentQuery.hoscode = hoscode
        val example = Example.of(departmentQuery)
        val departmentList = departmentRepository!!.findAll(example)
        val departmentMap = departmentList.stream().collect(
            Collectors.groupingBy(
                Function { obj: Department -> obj.bigcode })
        )
        for ((bigcode, departmentByBigCodeList) in departmentMap) {
            val departmentVo = DepartmentVo()
            departmentVo.depcode = bigcode
            departmentVo.depname = departmentByBigCodeList[0].bigname
            val children: MutableList<DepartmentVo> = ArrayList()
            for (department in departmentByBigCodeList) {
                val childDepartmentVo = DepartmentVo()
                childDepartmentVo.depcode = department.depcode
                childDepartmentVo.depname = department.depname
                children.add(childDepartmentVo)
            }
            departmentVo.children = children
            result.add(departmentVo)
        }
        return result
    }

    override fun getDepName(hoscode: String?, depcode: String?): String? {
        val department = departmentRepository!!.getDepartmentByHoscodeAndDepcode(hoscode, depcode)
        return department?.depname
    }

    override fun getDepartment(hoscode: String?, depcode: String?): Department? {
        return departmentRepository!!.getDepartmentByHoscodeAndDepcode(hoscode, depcode)
    }
}